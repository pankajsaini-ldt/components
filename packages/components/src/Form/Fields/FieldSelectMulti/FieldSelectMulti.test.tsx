/*

 MIT License

 Copyright (c) 2020 Looker Data Sciences, Inc.

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.

 */

import React from 'react'
import { fireEvent } from '@testing-library/react'
import { createWithTheme, renderWithTheme } from '@looker/components-test-utils'
import { FieldSelectMulti } from './FieldSelectMulti'

test('FieldSelectMulti supports labelWeight', () => {
  const { getByText } = renderWithTheme(
    <FieldSelectMulti
      label="Label"
      name="thumbsUp"
      id="thumbs-up"
      labelFontWeight="normal"
    />
  )
  const label = getByText('Label')
  expect(label).toHaveStyleRule('font-weight', '400')
})

test('Should accept a value', () => {
  const { getByText } = renderWithTheme(
    <FieldSelectMulti
      label="👍"
      name="thumbsUp"
      id="thumbs-up"
      values={['foobar']}
      options={[{ label: 'Foobar', value: 'foobar' }]}
    />
  )

  const chip = getByText('Foobar')
  expect(chip).toBeInTheDocument()
})

test('Should trigger onChange handler', () => {
  const handleChange = jest.fn()

  const { getByRole } = renderWithTheme(
    <FieldSelectMulti
      label="👍"
      name="thumbsUp"
      id="thumbs-up"
      values={['foobar']}
      onChange={handleChange}
      options={[{ label: 'Foobar', value: 'foobar' }]}
      closeOnSelect
    />
  )

  const input = getByRole('textbox')
  fireEvent.click(input) // open option list

  const option = getByRole('option')
  fireEvent.click(option) // click first option
  expect(handleChange).toHaveBeenCalledTimes(1)
})

test('A required FieldSelectMulti', () => {
  const component = createWithTheme(
    <FieldSelectMulti label="👍" name="thumbsUp" id="thumbs-up" required />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('A FieldSelectMulti with an error validation aligned to the bottom', () => {
  const { getByLabelText, getByText } = renderWithTheme(
    <FieldSelectMulti
      label="thumbs up"
      name="thumbsUp"
      id="thumbs-up"
      validationMessage={{ message: 'This is an error', type: 'error' }}
    />
  )
  expect(getByLabelText('thumbs up')).toBeVisible()
  expect(getByText('This is an error')).toBeVisible()
})

test('A FieldSelectMulti with description has proper aria setup', () => {
  const description = 'This is a description'

  const { container, getByRole } = renderWithTheme(
    <FieldSelectMulti
      id="test"
      defaultValues={['example']}
      description={description}
    />
  )

  const input = getByRole('input')
  const id = input.getAttribute('aria-describedby')
  expect(id).toBeDefined()

  const describedBy = container.querySelector(`#${id}`)
  expect(describedBy).toHaveTextContent(description)
})

test('A FieldText with error has proper aria setup', () => {
  const errorMessage = 'This is an error'

  const { container, getByDisplayValue } = renderWithTheme(
    <FieldSelectMulti
      id="test"
      defaultValues={['example']}
      validationMessage={{ message: errorMessage, type: 'error' }}
    />
  )

  const input = getByDisplayValue('example')
  const id = input.getAttribute('aria-describedby')
  expect(id).toBeDefined()

  const describedBy = container.querySelector(`#${id}`)
  expect(describedBy).toHaveTextContent(errorMessage)
})

test('A FieldSelectMulti with an error validation aligned to the right', () => {
  const component = createWithTheme(
    <FieldSelectMulti
      label="👍"
      name="thumbsUp"
      id="thumbs-up"
      validationMessage={{ message: 'This is an error', type: 'error' }}
    />
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
