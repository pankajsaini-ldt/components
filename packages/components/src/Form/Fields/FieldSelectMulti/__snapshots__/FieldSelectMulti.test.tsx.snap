// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A FieldSelectMulti with an error validation aligned to the right 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c6 {
  border: solid 1px;
  border-color: #DEE1E5;
  border-radius: 0.25rem;
  width: 100%;
  padding-right: 0.25rem;
  padding-top: 2px;
  padding-bottom: 2px;
  font-size: 0.875rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c10 {
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c17 {
  color: #707781;
  margin-right: 0.5rem;
  width: 18px;
  height: 18px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c15 {
  color: #CC1F36;
  margin-right: 0.25rem;
  width: 20px;
  height: 20px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c11 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 36px;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
}

.c11 input {
  border: none;
  background: transparent;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 0.875rem;
  height: 100%;
  width: 100%;
  outline: none;
  padding: 0;
}

.c11::-webkit-input-placeholder {
  color: #939BA5;
}

.c11::-moz-placeholder {
  color: #939BA5;
}

.c11:-ms-input-placeholder {
  color: #939BA5;
}

.c11::placeholder {
  color: #939BA5;
}

.c11:hover {
  border-color: #C1C6CC;
}

.c11:focus,
.c11:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c13 {
  border: solid 1px;
  border-color: #DEE1E5;
  border-radius: 0.25rem;
  height: 36px;
  width: 100%;
  padding-right: 0.75rem;
  padding-left: 0.75rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
  color: #343C42;
}

.c13:focus {
  outline: none;
}

.c2 {
  color: #343C42;
  font-size: 0.75rem;
  font-weight: 600;
}

.c20 {
  font-size: 0.75rem;
  color: #CC1F36;
}

.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  width: 100%;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: grid;
  grid-template-areas: "label detail" "input input" "messages messages";
}

.c0 .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  grid-area: input;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .c18 {
  grid-area: messages;
}

.c0 > .c1 {
  grid-area: label;
  line-height: 1rem;
  padding-bottom: 0.5rem;
}

.c0 .c22 {
  grid-area: detail;
  justify-self: end;
}

.c0 .c19 {
  margin-top: 0.5rem;
}

.c21 .c12 {
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  box-shadow: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: 100%;
  height: calc(36px - 6px);
}

.c21 .c12::-webkit-search-decoration,
.c21 .c12::-webkit-search-cancel-button,
.c21 .c12::-webkit-search-results-button,
.c21 .c12::-webkit-search-results-decoration {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c7 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
  border-color: #ED3B53;
}

.c7:hover {
  border-color: #C1C6CC;
}

.c7:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c7:hover {
  border-color: #CC1F36;
}

.c7:focus,
.c7:focus-within {
  border-color: #CC1F36;
  box-shadow: 0 0 0 2px #FFE5E9;
}

.c7 .c12 {
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  box-shadow: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: 100%;
  height: calc(36px - 6px);
}

.c7 .c12::-webkit-search-decoration,
.c7 .c12::-webkit-search-cancel-button,
.c7 .c12::-webkit-search-results-button,
.c7 .c12::-webkit-search-results-decoration {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c14 {
  display: grid;
  grid-auto-flow: column dense;
  grid-gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  max-height: 1.9rem;
}

.c16 {
  background: #DEE1E5;
  height: 70%;
  width: 1px;
}

.c8 {
  position: relative;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c8 .c5 {
  width: 100%;
}

.c8 .c12 {
  width: auto;
  min-width: 25%;
  padding-right: 0;
}

.c9 {
  padding-right: 0;
}

.c9 .c12 {
  cursor: default;
}

<div
  className="c0 "
>
  <label
    className="c1 c2"
    color="palette.charcoal700"
    fontSize="xsmall"
    fontWeight="semiBold"
    htmlFor="thumbs-up"
    id="thumbs-up-labeledby"
  >
    üëç
  </label>
  <div
    className="c3 "
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-owns="listbox-thumbs-up"
      className="c4 "
      display="flex"
      id="thumbs-up"
      name="thumbsUp"
      role="combobox"
    >
      <div
        aria-invalid={true}
        className="c5 c6 c7 c8 c9"
        fontSize="small"
        onClick={[Function]}
        onMouseDown={[Function]}
        width="100%"
      >
        <div
          className="c5 c10"
        >
          <div
            className="c11 c12 c13"
            onClick={[Function]}
          >
            <input
              aria-autocomplete="both"
              aria-describedby="thumbs-up-describedby"
              autoComplete="off"
              id="listbox-thumbs-up"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              readOnly={true}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="c14"
        >
          <div
            className="c15 "
            color="palette.red500"
            size={20}
          >
            <svg
              fill="currentColor"
              height="100%"
              viewBox="0 0 24 24"
              width="100%"
            >
              <title>
                Circle Info
              </title>
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
                fill="currentColor"
              />
            </svg>
          </div>
          <div
            className="c16"
          />
          <div
            className="c17 "
            color="palette.charcoal500"
            size={18}
          >
            <svg
              fill="currentColor"
              height="100%"
              viewBox="0 0 24 24"
              width="100%"
            >
              <title>
                Caret Down
              </title>
              <path
                d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c18 "
    id="thumbs-up-describedby"
  >
    <div
      className="c19 c20"
    >
      This is an error
    </div>
  </div>
</div>
`;

exports[`A required FieldSelectMulti 1`] = `
.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c8 {
  border: solid 1px;
  border-color: #DEE1E5;
  border-radius: 0.25rem;
  width: 100%;
  padding-right: 0.25rem;
  padding-top: 2px;
  padding-bottom: 2px;
  font-size: 0.875rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c12 {
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: start;
  align-content: flex-start;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c17 {
  color: #707781;
  margin-right: 0.5rem;
  width: 18px;
  height: 18px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.c13 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 36px;
  -webkit-box-pack: space-evenly;
  -webkit-justify-content: space-evenly;
  -ms-flex-pack: space-evenly;
  justify-content: space-evenly;
}

.c13 input {
  border: none;
  background: transparent;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 0.875rem;
  height: 100%;
  width: 100%;
  outline: none;
  padding: 0;
}

.c13::-webkit-input-placeholder {
  color: #939BA5;
}

.c13::-moz-placeholder {
  color: #939BA5;
}

.c13:-ms-input-placeholder {
  color: #939BA5;
}

.c13::placeholder {
  color: #939BA5;
}

.c13:hover {
  border-color: #C1C6CC;
}

.c13:focus,
.c13:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c15 {
  border: solid 1px;
  border-color: #DEE1E5;
  border-radius: 0.25rem;
  height: 36px;
  width: 100%;
  padding-right: 0.75rem;
  padding-left: 0.75rem;
  padding-top: 0rem;
  padding-bottom: 0rem;
  font-size: 0.875rem;
  color: #343C42;
}

.c15:focus {
  outline: none;
}

.c2 {
  color: #343C42;
  font-size: 0.75rem;
  font-weight: 600;
}

.c4 {
  position: absolute;
  height: 1px;
  width: 1px;
  overflow: hidden;
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
}

.c3::before {
  content: ' *';
  color: #CC1F36;
}

.c0 {
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  width: 100%;
  -webkit-align-items: left;
  -webkit-box-align: left;
  -ms-flex-align: left;
  align-items: left;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  display: grid;
  grid-template-areas: "label detail" "input input" "messages messages";
}

.c0 .c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  grid-area: input;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 .c18 {
  grid-area: messages;
}

.c0 > .c1 {
  grid-area: label;
  line-height: 1rem;
  padding-bottom: 0.5rem;
}

.c0 .c20 {
  grid-area: detail;
  justify-self: end;
}

.c0 .c19 {
  margin-top: 0.5rem;
}

.c9 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #FFFFFF;
}

.c9:hover {
  border-color: #C1C6CC;
}

.c9:focus-within {
  border-color: #9785F2;
  box-shadow: 0 0 0 2px #E8E5FF;
  outline: none;
}

.c9 .c14 {
  border: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  box-shadow: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  width: 100%;
  height: calc(36px - 6px);
}

.c9 .c14::-webkit-search-decoration,
.c9 .c14::-webkit-search-cancel-button,
.c9 .c14::-webkit-search-results-button,
.c9 .c14::-webkit-search-results-decoration {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c16 {
  display: grid;
  grid-auto-flow: column dense;
  grid-gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  justify-items: center;
  max-height: 1.9rem;
}

.c10 {
  position: relative;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c10 .c7 {
  width: 100%;
}

.c10 .c14 {
  width: auto;
  min-width: 25%;
  padding-right: 0;
}

.c11 {
  padding-right: 0;
}

.c11 .c14 {
  cursor: default;
}

<div
  className="c0 "
>
  <label
    className="c1 c2"
    color="palette.charcoal700"
    fontSize="xsmall"
    fontWeight="semiBold"
    htmlFor="thumbs-up"
    id="thumbs-up-labeledby"
  >
    üëç
    <span
      aria-hidden="true"
      className="c3"
    >
      <div
        className="c4"
      >
         required
      </div>
    </span>
  </label>
  <div
    className="c5 "
  >
    <div
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-owns="listbox-thumbs-up"
      className="c6 "
      display="flex"
      id="thumbs-up"
      name="thumbsUp"
      required={true}
      role="combobox"
    >
      <div
        aria-invalid={false}
        className="c7 c8 c9 c10 c11"
        fontSize="small"
        onClick={[Function]}
        onMouseDown={[Function]}
        width="100%"
      >
        <div
          className="c7 c12"
        >
          <div
            className="c13 c14 c15"
            onClick={[Function]}
          >
            <input
              aria-autocomplete="both"
              aria-describedby="thumbs-up-describedby"
              autoComplete="off"
              id="listbox-thumbs-up"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              readOnly={true}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="c16"
        >
          <div
            className="c17 "
            color="palette.charcoal500"
            size={18}
          >
            <svg
              fill="currentColor"
              height="100%"
              viewBox="0 0 24 24"
              width="100%"
            >
              <title>
                Caret Down
              </title>
              <path
                d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                fill="currentColor"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="c18 "
    id="thumbs-up-describedby"
  />
</div>
`;
